modules/platforms/cpp/core/src/impl/cluster/cluster_group_impl.cpp

```diff
@@ -48,6 +84,29 @@ namespace ignite
                 // No-op.
             }
 
+            SP_ClusterGroupImpl ClusterGroupImpl::ForAttribute(std::string name, std::string val)
+            {
+                common::concurrent::SharedPointer<interop::InteropMemory> mem = GetEnvironment().AllocateMemory();
+                interop::InteropOutputStream out(mem.Get());
+                binary::BinaryWriterImpl writer(&out, GetEnvironment().GetTypeManager());
+
+                writer.WriteString(name);
+                writer.WriteString(val);
+
+                out.Synchronize();
+
+                IgniteError err;
+                jobject target = InStreamOutObject(Command::FOR_ATTRIBUTE, *mem.Get(), err);
+                IgniteError::ThrowIfNeeded(err);
+
+                return SP_ClusterGroupImpl(new ClusterGroupImpl(GetEnvironmentPointer(), target));
+            }
+
+            SP_ClusterGroupImpl ClusterGroupImpl::ForDataNodes(std::string cacheName)
+            {
+                return ForCacheNodes(cacheName, Command::FOR_DATA);
+            }
+
             SP_ClusterGroupImpl ClusterGroupImpl::ForServers()
```

Comment:
Check this method please - it may really work as "ForCpp" for historical reasons. Now we keep lazy-initialized value in Ignite which this method returns. Review it - maybe we need to change this logic.